tracing_test.go

```diff
@@ -21,314 +21,531 @@
+		downstreamReq := &TracingRequest{ForwardCount: req.ForwardCount - 1}
+		childResp = downstream(ctx, downstreamReq)
 	}
 
-	return &TracingResponse{
-		TraceID:        span.TraceID(),
-		SpanID:         span.SpanID(),
-		ParentID:       span.ParentID(),
-		TracingEnabled: span.TracingEnabled(),
-		Child:          childResp,
-	}, nil
+	resp := &TracingResponse{Child: childResp}
+	resp.observeSpan(ctx)
+
+	return resp, nil
 }
 
-func (h *traceHandler) onError(ctx context.Context, err error) {
-	h.t.Errorf("onError %v", err)
+// RawHandler tests tracing over Raw encoding
+type RawHandler struct {
+	traceHandler
 }
 
-func TestTracingPropagates(t *testing.T) {
-	WithVerifiedServer(t, nil, func(ch *Channel, hostPort string) {
-		handler := &traceHandler{t: t, ch: ch}
-		json.Register(ch, json.Handlers{
-			"call": handler.call,
-		}, handler.onError)
+func (h *RawHandler) Handle(ctx context.Context, args *raw.Args) (*raw.Res, error) {
+	req := new(TracingRequest).fromRaw(args)
+	res, err := h.handleCall(ctx, req,
+		func(ctx context.Context, req *TracingRequest) *TracingResponse {
+			_, arg3, _, err := raw.Call(ctx, h.ch, h.ch.PeerInfo().HostPort,
+				h.ch.PeerInfo().ServiceName, "rawcall", nil, req.toRaw())
+			require.NoError(h.t, err)
+			return new(TracingResponse).fromRaw(h.t, arg3)
+		})
+	require.NoError(h.t, err)
+	return res.toRaw(h.t), nil
+}
 
-		ctx, cancel := json.NewContext(time.Second)
-		defer cancel()
+func (h *RawHandler) OnError(ctx context.Context, err error) { h.t.Errorf("onError %v", err) }
 
-		peer := ch.Peers().GetOrAdd(ch.PeerInfo().HostPort)
-		var response TracingResponse
-		require.NoError(t, json.CallPeer(ctx, peer, ch.PeerInfo().ServiceName, "call", &TracingRequest{
-			ForwardCount: 1,
-		}, &response))
-
-		clientSpan := CurrentSpan(ctx)
-		require.NotNil(t, clientSpan)
-		assert.Equal(t, uint64(0), clientSpan.ParentID())
-		assert.NotEqual(t, uint64(0), clientSpan.TraceID())
-		assert.True(t, clientSpan.TracingEnabled(), "Tracing should be enabled")
-		assert.Equal(t, clientSpan.TraceID(), response.TraceID)
-		assert.Equal(t, clientSpan.SpanID(), response.ParentID)
-		assert.True(t, response.TracingEnabled, "Tracing should be enabled")
-		assert.Equal(t, response.TraceID, response.SpanID, "traceID = spanID for root span")
-
-		nestedResponse := response.Child
-		require.NotNil(t, nestedResponse)
-		assert.Equal(t, clientSpan.TraceID(), nestedResponse.TraceID)
-		assert.Equal(t, response.SpanID, nestedResponse.ParentID)
-		assert.True(t, response.TracingEnabled, "Tracing should be enabled")
-		assert.NotEqual(t, response.SpanID, nestedResponse.SpanID)
-	})
+// JSONHandler tests tracing over JSON encoding
+type JSONHandler struct {
+	traceHandler
 }
 
-func TestTraceReportingEnabled(t *testing.T) {
-	initialTime := time.Date(2015, 2, 1, 10, 10, 0, 0, time.UTC)
+func (h *JSONHandler) callJSON(ctx json.Context, req *TracingRequest) (*TracingResponse, error) {
```

Comment:
can we move json tests into the `json` package. if you want to share code, feel free to add a `tchannel/testutils/tracetest` package with helpful utilities for both, but this package should not be testing the json package.